java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java

```diff
@@ -0,0 +1,411 @@
+        return to.setConstraint(programState, shouldNotInverse().equals(booleanConstraint) ? boolConstraint : boolConstraint.inverse());
+      } else if (constraintLeft instanceof NullConstraint) {
+        NullConstraint nullConstraint = (NullConstraint) constraintLeft;
+        if (nullConstraint.equals(NullConstraint.NULL)) {
+          return to.setConstraint(programState, shouldNotInverse().equals(booleanConstraint) ? nullConstraint : nullConstraint.inverse());
+        } else if (shouldNotInverse().equals(booleanConstraint)) {
+          return to.setConstraint(programState, nullConstraint);
+        }
+      }
+      return ImmutableList.of(programState);
+    }
+
+  }
+  class NotEqualToSymbolicValue extends BinarySymbolicValue {
+
+    public NotEqualToSymbolicValue(int id) {
+      super(id);
+    }
+
+    @Override
+    public String toString() {
+      return "NEQ_TO_" + super.toString();
+    }
+
+    @Override
+    BooleanConstraint shouldNotInverse() {
+      return BooleanConstraint.FALSE;
+    }
+  }
+  class EqualToSymbolicValue extends BinarySymbolicValue {
+
+    public EqualToSymbolicValue(int id) {
+      super(id);
+    }
+
+    @Override
+    BooleanConstraint shouldNotInverse() {
+      return BooleanConstraint.TRUE;
+    }
+
+  }
+
+  abstract class UnarySymbolicValue extends ObjectSymbolicValue {
+    protected SymbolicValue operand;
+
+    public UnarySymbolicValue(int id) {
+      super(id);
+    }
+
+    @Override
+    public void computedFrom(List<SymbolicValue> symbolicValues) {
+      Preconditions.checkArgument(symbolicValues.size() == 1);
+      this.operand = symbolicValues.get(0);
+    }
+
+  }
+
+  class NotSymbolicValue extends UnarySymbolicValue {
+
+    public NotSymbolicValue(int id) {
+      super(id);
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
+      return operand.setConstraint(programState, booleanConstraint.inverse());
+    }
+  }
+
+  class InstanceOfSymbolicValue extends UnarySymbolicValue {
+    public InstanceOfSymbolicValue(int id) {
+      super(id);
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
+      if (BooleanConstraint.TRUE.equals(booleanConstraint)) {
+        if (NullConstraint.NULL.equals(programState.constraints.get(operand))) {
+          // irrealizable constraint : instance of true if operand is null
+          return ImmutableList.of();
+        }
+        // if instanceof is true then we know for sure that expression is not null.
+        List<ProgramState> ps = operand.setConstraint(programState, NullConstraint.NOT_NULL);
+        if (ps.size() == 1 && ps.get(0).equals(programState)) {
+          // FIXME we already know that operand is NOT NULL, so we add a different constraint to distinguish program state. Typed Constraint
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)